client/src/worker_registry/mod.rs

```diff
@@ -0,0 +1,288 @@
+//! This module enables the tracking of workers that are associated with a client instance.
+//! This is needed to implement Eager Workflow Start, a latency optimization in which the client,
+//!  after reserving a slot, directly forwards a WFT to a local worker.
+
+use rand::{seq::SliceRandom, thread_rng};
+use std::{collections::HashMap, fmt::Debug};
+use temporal_sdk_core_protos::temporal::api::workflowservice::v1::PollWorkflowTaskQueueResponse;
+
+/// This trait wraps a worker, which provides WFT processing slots.
+#[cfg_attr(test, mockall::automock)]
+pub trait SlotProvider: Send + Sync + Debug {
+    /// A unique identifier for the worker.
+    fn uuid(&self) -> String;
+    /// The namespace for the WFTs that it can process.
+    fn namespace(&self) -> String;
+    /// The task queue this provider listens to.
+    fn task_queue(&self) -> String;
+    /// Try to reserve a slot with matching namespace and task queue.
+    fn try_reserve_wft_slot(&self) -> Option<Box<dyn Slot>>;
+}
+
+/// This trait represents a slot reserved for processing a WFT by a worker.
+#[cfg_attr(test, mockall::automock)]
+pub trait Slot: Send + Sync {
+    /// Consumes this slot by dispatching a WFT to its worker. This can only be called once.
+    fn schedule_wft(&mut self, task: PollWorkflowTaskQueueResponse) -> Result<(), anyhow::Error>;
+}
+
+/// This trait enables local workers to made themselves visible to a shared client instance.
+pub trait WorkerRegistry: Send + Sync {
+    /// Register a local worker that can provide WFT processing slots.
+    fn register(&mut self, provider: Box<dyn SlotProvider>);
+    /// Unregister a provider, typically when its worker starts shutdown.
+    fn unregister(&mut self, uuid: String);
+}
+
+/// Implements a [WorkerRegistry] and provides a convenient method
+/// to find compatible slots within the collection.
+#[derive(Default, Debug)]
+pub struct SlotManager {
+    /// Maps keys, i.e., namespace#task_queue, to providers.
+    providers: HashMap<String, Vec<Box<dyn SlotProvider>>>,
+    /// Maps uuids to keys in `providers`.
+    index: HashMap<String, String>,
+}
+
+fn to_key(namespace: &str, task_queue: &str) -> String {
+    format!("{}#{}", namespace, task_queue)
+}
```

Comment:
No need to make up this format string. Just make a type which has `namespace` and `task_queue` fields.

Reply:
Happy to do that, as I just replied above, it is also used as a value in another hashmap, and that's why I did that...

Reply:
Done, much cleaner with a tuple key and all the macros help...